### DISCUSSION TITLE
What allows the MorganFingerprint to distinguish these two molecules?

### DISCUSSION THREAD

**Discussion by JanoschMenke** [rdkit/rdkit#5066] [Q&A]
Hi Guys,

so I am just playing around the ECFP4 and their limitations. And I come across these two molecules, which can not be distinguished by the ECF4/MorganFingerprint(radius=2).  Similarity of 1.

```
mol1 = Chem.MolFromSmiles("C1CCN(C1)N1CCCC1")
mol2 = Chem.MolFromSmiles("C1CCN2CCCCN2C1")
fp1 = Chem.GetMorganFingerprint(mol1,2)
fp2 = Chem.GetMorganFingerprint(mol2,2)
Draw.MolsToGridImage([mol1, mol2],molsPerRow=2)
```
When you use a radius of 3 then the two Molecules can be distinguished (similarity of 0.87):

```
bit_dict = {}
fp1 = Chem.GetMorganFingerprintAsBitVect(mol1, radius=3, bitInfo=bit_dict)
bit_dict

>>>{158: ((3, 3), (5, 3)),
 204: ((6, 3), (2, 3)),
 385: ((3, 1), (5, 1)),
 926: ((0, 0), (1, 0), (2, 0), (4, 0), (6, 0), (7, 0), (8, 0), (9, 0)),
 935: ((3, 0), (5, 0)),
 1028: ((1, 1), (7, 1), (8, 1), (0, 1)),
 1478: ((5, 2), (3, 2)),
 1480: ((2, 1), (6, 1), (4, 1), (9, 1)),
 1632: ((0, 3), (7, 3)),
 1843: ((2, 2), (4, 2), (6, 2), (9, 2)),
 1998: ((7, 2), (1, 2), (0, 2), (8, 2))}
```

```
bit_dict = {}
fp1 = Chem.GetMorganFingerprintAsBitVect(mol2, radius=3, bitInfo=bit_dict)
bit_dict

>>> {158: ((3, 3),),
 204: ((4, 3), (2, 3), (9, 3), (7, 3)),
 385: ((3, 1), (8, 1)),
 926: ((0, 0), (1, 0), (2, 0), (4, 0), (5, 0), (6, 0), (7, 0), (9, 0)),
 935: ((3, 0), (8, 0)),
 1028: ((1, 1), (5, 1), (6, 1), (0, 1)),
 1478: ((3, 2), (8, 2)),
 1480: ((2, 1), (4, 1), (7, 1), (9, 1)),
 1632: ((5, 3), (6, 3), (1, 3), (0, 3)),
 1843: ((2, 2), (4, 2), (7, 2), (9, 2)),
 1998: ((5, 2), (6, 2), (1, 2), (0, 2))}
```
I do not understand how that is possible. Also the bits appear to be same. I assume it has something to do with the bonds, as based on the atoms features alone the distinction should be impossible (Based on my understanding of the generating Algorithm).

I know it is more an theoretic rather the code based questions, but maybe you will find the time to answer.

In any case thanks for the awesome package:)

**Comment by greglandrum** [comment]
Hi @JanoschMenke
First a minor correction: both of those molecules set the same bits, even at radius 3. Look at the keys from the dictionary and you'll see that. So for the bit vector fingerprints you are calculating here the similarity will be 1.0

The difference is how many atom environments are found, and the differences here are due to the fact that the Morgan fingeprinting algorithm removes redundant atom environments - atom environments which contain exactly the same atoms and bonds. This is part of the original published ECFP paper.

In the case of your molecules:
![image](https://user-images.githubusercontent.com/540511/157590442-328e592c-f240-437b-80ff-ad9cd37c0b14.png)

The environment of radius three starting at either atom 2 or atom 4 in mol1 includes atoms (2,1,3,0,4,5,6,9) and all of the bonds between them. Since these are redundant only one is generated.

The equivalent environments for mol2, centered at atoms 4 and 7, include different atoms:
- atom 4: (4,5,3,6,2,8,7,1,9)
- atom 7: (7,6,8,5,3,9,4,0,2)
so they do not end up being marked as redundant

If you like, you can turn this off by telling the RDKit to generate redundant atom environments:
```
bi1 = {}
fp1 = rdMolDescriptors.GetMorganFingerprintAsBitVect(mol1, radius=3, bitInfo=bi1, includeRedundantEnvironments=True )
bi2 = {}
fp2 = rdMolDescriptors.GetMorganFingerprintAsBitVect(mol2, radius=3, bitInfo=bi2, includeRedundantEnvironments=True)
```
produces:
```
>>> bi1
{158: ((3, 3), (5, 3)),
 204: ((6, 3), (9, 3), (2, 3), (4, 3)),
 385: ((3, 1), (5, 1)),
 926: ((0, 0), (1, 0), (2, 0), (4, 0), (6, 0), (7, 0), (8, 0), (9, 0)),
 935: ((3, 0), (5, 0)),
 1028: ((1, 1), (7, 1), (8, 1), (0, 1)),
 1478: ((5, 2), (3, 2)),
 1480: ((2, 1), (6, 1), (4, 1), (9, 1)),
 1632: ((0, 3), (1, 3), (7, 3), (8, 3)),
 1843: ((2, 2), (4, 2), (6, 2), (9, 2)),
 1998: ((7, 2), (1, 2), (0, 2), (8, 2))}
>>> bi2
{158: ((3, 3), (8, 3)),
 204: ((4, 3), (2, 3), (9, 3), (7, 3)),
 385: ((3, 1), (8, 1)),
 926: ((0, 0), (1, 0), (2, 0), (4, 0), (5, 0), (6, 0), (7, 0), (9, 0)),
 935: ((3, 0), (8, 0)),
 1028: ((1, 1), (5, 1), (6, 1), (0, 1)),
 1478: ((3, 2), (8, 2)),
 1480: ((2, 1), (4, 1), (7, 1), (9, 1)),
 1632: ((5, 3), (6, 3), (1, 3), (0, 3)),
 1843: ((2, 2), (4, 2), (7, 2), (9, 2)),
 1998: ((5, 2), (6, 2), (1, 2), (0, 2))}
```

**Comment by JanoschMenke** [comment]
Hi Greg,

thank you for your detailed response. You are right, that I made a mistake when talking about the similarity. 

When analyzing the similarity without folding the fingerprint:
```
fp1 = Chem.GetMorganFingerprint(mol1,3)
fp2 = Chem.GetMorganFingerprint(mol2,3)
DataStructs.TanimotoSimilarity(fp1, fp2)
```
 would result in a similarity of 0.855.
 
However, as you stated, when comparing the folded bit vectors, the similarity is indeed 1. 
I was confused by this difference, but your explanation made it clear to me.

I was aware of the fact that the ECFP removes redundant structures. But I always thought that: 

> exactly the same atoms and bonds

would just refer to atoms with the same features/properties, not the  exact atom. So in my mind 
- atom 4: (4,5,3,6,2,8,7,1,9)
- atom 7: (7,6,8,5,3,9,4,0,2)
would also be considered the exact same structures. And one would have been removed.

As I understand it now, the actual atom ID is also considered in the MorganFingerprint, right? I always that this is not included during the algorithm, which explain why I did not follow the grasp the difference in similarity
